dags/synapse-dataset-to-croissant.py

```diff
@@ -0,0 +1,563 @@
+            FROM 
+                SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST,
+            LATERAL FLATTEN(input => ITEMS) ITEM
+            WHERE
+                node_type = 'datasetcollection'
+                AND id = {dataset_collection_without_syn}
+            """
+            snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+            ctx = snow_hook.get_conn()
+            cs = ctx.cursor()
+
+            otel_logger.info(
+                f"Performing query for dataset_collection {dataset_collection}")
+
+            try:
+                cs.execute(query)
+                datasets = cs.fetch_pandas_all()
+                cs.close()
+                results = []
+                for syn_id in datasets["DATASET_IDS"].tolist():
+                    results.append({"dataset_id": syn_id.replace("syn", "")})
+                results
+            except Exception as ex:
+                otel_logger.exception(
+                    "Failed to query snowflake for datasets.")
+                otel_tracer.span_processor.force_flush()
+                otel_logger.handlers[0].flush()
+                raise ex
+
+        otel_tracer.span_processor.force_flush()
+        otel_logger.handlers[0].flush()
+        return results
+
+    @task
+    def combine_dataset_lists(root_carrier_context: Dict, dataset_ids: List[List[Dict[str, str]]]) -> List[Dict[str, str]]:
+        """
+        Combine the dataset IDs into a single list. This is used to force the
+        tasks to join together before being expanded again.
+
+        Arguments:
+            dataset_ids: The list of dataset IDs to combine.
+
+        Returns:
+            The combined list of dataset IDs.
+        """
+        with otel_tracer.start_as_current_span("combine_dataset_lists", context=TraceContextTextMapPropagator().extract(root_carrier_context)) as span:
+            return_value = [
+                dataset for datasets in dataset_ids for dataset in datasets]
+        otel_tracer.span_processor.force_flush()
+        return return_value
+
+    @task
+    def query_snowflake_and_push_croissant_file(root_carrier_context: Dict, dataset_id: str, **context) -> None:
+        """
+        Query the Snowflake database for the dataset in order to perform the Croissant
+        transformation and push the results to S3.
+
+        Arguments:
+            dataset_id: The ID of the dataset to query in Snowflake.
+        """
+        with otel_tracer.start_as_current_span("query_snowflake_and_push_croissant_file", context=TraceContextTextMapPropagator().extract(root_carrier_context)) as span:
+            if not dataset_id:
+                otel_logger.info("No dataset found")
+                otel_tracer.span_processor.force_flush()
+                otel_logger.handlers[0].flush()
+                return
+
+            span.set_attribute("airflow.dataset_id", f"syn{dataset_id}")
+
+            query = f"""
```

Comment:
I wonder as we continue performing data operations with Airflow + Snowflake if there is a better pattern for storing/using queries that we should consider. This one is pretty complex and having it just be a long string block within Python code feels kind of brittle.

Reply:
We could abstract recurring queries/CTEs into [user-defined functions](https://docs.snowflake.com/en/sql-reference/sql/create-function) or stored procedures, although it wouldn't be much help for queries with very specific use-cases like this one, and I'm not a fan of abstracting things if it makes the query harder to debug. But still this is a good point and I have an [exploration ticket](https://sagebionetworks.jira.com/browse/SNOW-234) on this where we can add more robust DAGs as a motivator.

Reply:
Thanks for the comments. To me, it seems to come down to whether we need to be able to run the queries outside the context on this DAG.

At the moment I do not see that need, however, based on Jenny's ticket we could start to define the standard for introducing snowflake functions in a standardized way. At that point I think we could then port the queries over and refactor the DAGs to call them.